alf/utils/image_networks.py

```diff
@@ -0,0 +1,216 @@
+                or the final FC output size
+
+        Returns:
+            a tuple representing the output shape
+        """
+        if self._postprocess_fc_layers and not conv:
+            return (self._postprocess_fc_layer_params[-1], )
+
+        size = input_size
+        for filters, kernel_size, strides in self._conv_layer_params:
+            if self._padding == "same":
+                size = (size - 1) // strides + 1
+            else:
+                size = (size - kernel_size) // strides + 1
+        return (size, size, filters)
+
+    def call(self, inputs):
+        z = inputs
+        for conv_l in self._conv_layers:
+            z = conv_l(z)
+        if self._postprocess_fc_layers:
+            z = tf.reshape(z, [tf.shape(z)[0], -1])
+            for fc_l in self._postprocess_fc_layers:
+                z = fc_l(z)
+        return z
+
+
+@gin.configurable
+class ImageDecodingNetwork(tf.keras.models.Model):
+    """
+    A general template class for create transposed convolutional decoding networks.
+    """
+
+    def __init__(self,
+                 deconv_layer_params,
+                 start_decoding_size,
+                 start_decoding_filters,
+                 padding="same",
+                 preprocess_fc_layer_params=None,
+                 activation_fn=tf.nn.relu,
+                 output_activation_fn=tf.nn.tanh,
+                 name="image_decoding_network"):
+        """
+        Initialize the layers for decoding a latent vector into an image.
+
+        Args:
+            deconv_layer_params (list[tuple]): a non-empty list of elements
+                (num_filters, kernel_size, strides).
+            start_decoding_size (int): the initial size we'd like to have for
+                the feature map
+            start_decoding_filters (int): the initial number of fitlers we'd like
+                to have for the feature map. Note that given this value and
+                `start_decoding_size`, we always first project an input latent
+                vector into a vector of an appropriate length so that it can be
+                reshaped into (`start_decoding_size`, `start_decoding_size`,
+                `start_decoding_filters`).
+            padding (str): "same" or "valid", see tf.keras.layers.Conv2DTranspose
+            preprocess_fc_layer_params (tuple[int]): a list of fc layer units.
+                These fc layers are used for preprocessing the latent vector before
+                transposed convolutions.
+            activation_fn (tf.nn.activation): activation for hidden layers
+            output_activation_fn (tf.nn.activation): activation for the output
+                layer. Usually our image inputs are normalized to [0, 1] or [-1, 1],
+                so this function should be tf.nn.sigmoid or tf.nn.tanh.
+            name (str): network name
+        """
+        super().__init__(name=name)
+
+        assert isinstance(deconv_layer_params, list)
```

Comment:
This assertion seems to be inconsistent with the comment on support tuple as well ("deconv_layer_params (list[tuple])").